### PR TITLE
Robustness fix patch txn onfailure

### PR DESCRIPTION
**Pull request by serathius**
cc @ahrtr @ptabor @jmhbnz @chaochn47 @fuweid

### CODE REVIEW

File: tests/robustness/validate/patch_history.go [hunk 1]
```diff
@@ -84,17 +110,25 @@ func matchWatchEvent(request *model.TxnRequest, watchEvents map[model.Event]traf
 	return nil
 }
 
-func hasNonUniqueWriteOperation(request *model.TxnRequest) bool {
-	for _, etcdOp := range request.OperationsOnSuccess {
+func canBeDiscarded(request *model.TxnRequest) bool {
+	return operationsCanBeDiscarded(request.OperationsOnSuccess) && operationsCanBeDiscarded(request.OperationsOnFailure)
```
Comment (chaochn47) [inline tests/robustness/validate/patch_history.go:114]: Is the motivation that porcupine.Operation(s) that contains `Delete and Range Type OnFailure operations` should have been discarded but not causing the robustness test timeout?

File: tests/robustness/validate/patch_history_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		},
+		{
+			name: "failed txn remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{}, []clientv3.Op{}, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: false,
```
Comment (chaochn47) [inline tests/robustness/validate/patch_history_test.go:144]: Maybe rename the test case name to "non put/delete failed txn is dropped"?

File: tests/robustness/validate/patch_history_test.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		{
+			name: "failed txn put&delete/put&delete is dropped",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{clientv3.OpPut("key", "value1"), clientv3.OpDelete("key")}, []clientv3.Op{clientv3.OpPut("key", "value2"), clientv3.OpDelete("key")}, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: false,
+		},
```
Comment (chaochn47) [inline tests/robustness/validate/patch_history_test.go:185]: For other reviewers, that's the most important one matched the above statement. 

> Currently etcd nor Kubernetes traffic code doesn't generate such transactions.

**Review by chaochn47 — APPROVED** [review]
LGTM with one minor comment. Thanks @serathius!

**Comment by chaochn47** [comment]
```=== FAIL: integration/clientv3/examples ExampleCluster_memberAddAsLearner``` can be ignored given it's being worked on to be fixed.

File: tests/robustness/validate/patch_history.go (hunk 1 above)
Comment (chaochn47) [reply tests/robustness/validate/patch_history.go:115]: Thanks for clarifying.

File: tests/robustness/validate/patch_history_test.go (hunk 2 above)
Comment (serathius) [reply tests/robustness/validate/patch_history_test.go:144]: Doesn't match the previously established pattern. We could rename it to `failed empty txn remains`

File: tests/robustness/validate/patch_history_test.go (hunk 3 above)
Comment (serathius) [reply tests/robustness/validate/patch_history_test.go:185]: Not accurate, this also applies to `failed txn put/delete remains`, `failed txn delete/put remains` and many other 
test cases.

**Pull request merged**